RULES/hdl/xsim.go

```diff
@@ -2,74 +2,495 @@ package hdl
+
+	// Default flag values
+	vsim_flags := " --onfinish quit --log " + log_file.String()
+	seed_flag := " --sv_seed 1"
+	verbosity_flag := " --testplusarg verbosity=DVM_VERB_NONE"
+	mode_flag := ""
+	plusargs_flag := ""
+
+	// Collect do-files here
+	var do_flags []string
+
+	// Turn off output unless verbosity is activated
+	print_output := false
+
+	// Parse additional arguments
+	for _, arg := range args {
+		if strings.HasPrefix(arg, "-seed=") {
+			// Define simulator seed
+			var seed int64
+			if _, err := fmt.Sscanf(arg, "-seed=%d", &seed); err == nil {
+				seed_flag = fmt.Sprintf(" --sv_seed %d", seed)
+			} else {
+				log.Fatal("-seed expects an integer argument!")
+			}
+		} else if strings.HasPrefix(arg, "-verbosity=") {
+			// Define verbosity level
+			var level string
+			if _, err := fmt.Sscanf(arg, "-verbosity=%s", &level); err == nil {
+				verbosity_flag, print_output = xsimVerbosityLevelToFlag(level)
+			} else {
+				log.Fatal("-verbosity expects an argument of 'low', 'medium', 'high' or 'none'!")
+			}
+		} else if strings.HasPrefix(arg, "+") {
+			// All '+' arguments go directly to the simulator
+			plusargs_flag = plusargs_flag + " --testplusarg " + strings.TrimPrefix(arg, "+")
+		}
+	}
+
+	// Create optional command preamble
+	cmd_preamble, testcase = xsimPreamble(rule, testcase)
+
+	cmd_echo := ""
+	if rule.Params != nil && params != "" {
+		// Update coverage database name based on parameters. We cannot merge
+		// different parameter sets, do we have to make a dedicated main database
+		// for this parameter set.
+		cmd_echo = "Testcase " + params
+
+		// Update with testcase if specified
+		if testcase != "" {
+			cmd_echo = cmd_echo + "/" + testcase + ":"
+			testcase = params + "_" + testcase
+		} else {
+			cmd_echo = cmd_echo + ":"
+			testcase = params
+		}
+	} else {
+		// Update coverage database name with testcase alone, main database stays
+		// the same
+		if testcase != "" {
+			cmd_echo = "Testcase " + testcase + ":"
+		} else {
+			testcase = "default"
+		}
+	}
+
+	cmd_postamble := ""
+	if gui {
+		mode_flag = " --gui"
+		if rule.WaveformInit != nil {
+			do_flags = append(do_flags, rule.WaveformInit.String())
+		}
+	} else {
+		mode_flag = " --runall"
+		cmd_newline := ":"
+		if cmd_echo != "" {
+			cmd_newline = "echo"
+		}
+
+		cmd_postamble = fmt.Sprintf("|| { %s; cat %s; exit 1; }", cmd_newline, log_file.String())
+	}
+
+	vsim_flags = vsim_flags + mode_flag + seed_flag +
+		verbosity_flag + plusargs_flag + XsimFlags.Value()
+
+	for _, do_flag := range do_flags {
+		vsim_flags = vsim_flags + " --tclbatch " + do_flag
+	}
+
+	// Using this part of the command we send the stdout into a black hole to
+	// keep the output clean
+	cmd_devnull := ""
+	if !print_output {
+		cmd_devnull = "> /dev/null"
+	}
+
+	cmd := fmt.Sprintf("{ echo -n %s && xsim %s %s %s && "+
+		"{ { ! grep -q FAILURE %s; } && echo PASS; } }",
+		cmd_echo, vsim_flags, rule.Name+params, cmd_devnull, log_file.String())
```

Comment:
I believe this only works reliably with our new testing environment. But that's not yet implemented consistently.